apps/rush-lib/src/pluginFramework/PluginLoader/PluginLoaderBase.ts

```diff
@@ -0,0 +1,227 @@
+  ITerminal,
+  JsonFile,
+  JsonObject,
+  JsonSchema
+} from '@rushstack/node-core-library';
+import * as path from 'path';
+import { CommandLineConfiguration } from '../../api/CommandLineConfiguration';
+import { RushConfiguration } from '../../api/RushConfiguration';
+import { IRushPluginConfigurationBase } from '../../api/RushPluginsConfiguration';
+import { RushConstants } from '../../logic/RushConstants';
+import { IRushPlugin } from '../IRushPlugin';
+
+export interface IRushPluginManifest {
+  pluginName: string;
+  description: string;
+  entryPoint?: string;
+  optionsSchema?: string;
+  associatedCommands?: string[];
+  commandLineJsonFilePath?: string;
+}
+
+export interface IRushPluginManifestJson {
+  plugins: IRushPluginManifest[];
+}
+
+export interface IPluginLoaderOptions {
+  pluginConfiguration: IRushPluginConfigurationBase;
+  rushConfiguration: RushConfiguration;
+  terminal: ITerminal;
+}
+
+export class PluginLoaderBase {
+  protected static _jsonSchema: JsonSchema = JsonSchema.fromFile(
+    path.join(__dirname, '../../schemas/rush-plugin-manifest.schema.json')
+  );
+
+  protected _rushConfiguration: RushConfiguration;
+  protected _packageName: Readonly<string>;
+  protected _pluginName: Readonly<string>;
+  protected _terminal: ITerminal;
+  protected _manifestCache!: Readonly<IRushPluginManifest>;
+
+  public constructor({ pluginConfiguration, rushConfiguration, terminal }: IPluginLoaderOptions) {
+    this._packageName = pluginConfiguration.packageName;
+    this._pluginName = pluginConfiguration.pluginName;
+    this._rushConfiguration = rushConfiguration;
+    this._terminal = terminal;
+  }
+
+  public load(): IRushPlugin | undefined {
+    const resolvedPluginPath: string | undefined = this._resolvePlugin();
+    if (!resolvedPluginPath) {
+      return undefined;
+    }
+    const pluginOptions: JsonObject = this._getPluginOptions();
+
+    return this._loadAndValidatePluginPackage(resolvedPluginPath, pluginOptions);
+  }
+
+  public get packageName(): string {
+    return this._packageName;
+  }
+
+  public get pluginName(): string {
+    return this._pluginName;
+  }
+
+  public get pluginManifest(): IRushPluginManifest {
+    return this._getRushPluginManifest();
+  }
+
+  public getPackageFolder(): string {
+    const packageFolder: string = Import.resolvePackage({
+      baseFolderPath: __dirname,
+      packageName: this._packageName
+    });
+    return packageFolder;
+  }
+
+  public getCommandLineConfiguration(): CommandLineConfiguration | undefined {
+    const commandLineJsonFilePath: string | undefined = this._getCommandLineJsonFilePath();
+    if (!commandLineJsonFilePath || !FileSystem.exists(commandLineJsonFilePath)) {
+      return undefined;
+    }
+    const commandLineConfiguration: CommandLineConfiguration =
+      CommandLineConfiguration.loadFromFileOrDefault(commandLineJsonFilePath);
+    for (const additionalPathFolder of this._getCommandLineAdditionalPathFolders().reverse()) {
```

Comment:
@dmichon-msft  Implementers return an array like `['example1', 'example2', 'example3']`.  This order makes sense, because it needs to end up in the shell environment as `PATH=example1;example2;example3;C:\Windows;...`.

He's using `reverse()`  here because `commandLineConfiguration.prependAdditionalPathFolder(additionalPathFolder);` adds one thing to the front (e.g. `PATH=example3;C:\Windows;...`), so if we do this in a loop then we need to go in reverse order.